File: test_full_audit.cpp
// Full C++ audit: loaded-params transforms, Jacobi fit, whitening variance,
// transform_one / transform_one_raw consistency, error handling.
#include <cassert>
#include <cmath>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#include "pca.hpp"
#include "lcg_full_pca_params.hpp"   // full_pca_params::make_pca()

// ---------------------------------------------------------------------------
// LCG + data generation (must match test_full_audit.py exactly)
// ---------------------------------------------------------------------------
static constexpr uint32_t LCG_SEED = 143;
static constexpr int      GEN_N    = 3000;
static constexpr int      GEN_D    = 100;
static constexpr int      REF_NTX  = 50;
static constexpr int      REF_K    = 20;

static inline uint32_t lcg(uint32_t seed) {
    return 1664525u * seed + 1013904223u;
}

static std::vector<std::vector<double>> gen_data() {
    std::vector<std::vector<double>> X(GEN_N, std::vector<double>(GEN_D));
    uint32_t seed = LCG_SEED;
    for (int i = 0; i < GEN_N; ++i)
        for (int j = 0; j < GEN_D; ++j) {
            seed = lcg(seed);
            X[i][j] = seed / 4294967296.0;
        }
    return X;
}

static void center_inplace(std::vector<std::vector<double>>& X) {
    const int N = static_cast<int>(X.size());
    const int D = static_cast<int>(X[0].size());
    std::vector<double> mean(D, 0.0);
    for (const auto& row : X)
        for (int d = 0; d < D; ++d)
            mean[d] += row[d];
    for (auto& m : mean) m /= N;
    for (auto& row : X)
        for (int d = 0; d < D; ++d)
            row[d] -= mean[d];
}

// ---------------------------------------------------------------------------
// Reference file reader
// ---------------------------------------------------------------------------
struct Ref {
    std::vector<std::vector<double>> tx;    // (N_TX, K)
    std::vector<std::vector<double>> wtx;   // (N_TX, K)
    std::vector<double>              evals; // (K,)
};

static Ref read_ref(const std::string& path) {
    std::ifstream f(path);
    if (!f) throw std::runtime_error("Cannot open " + path);
    Ref r;
    auto read_block2d = [&](std::vector<std::vector<double>>& blk) {
        int rows, cols;
        f >> rows >> cols;
        blk.resize(rows, std::vector<double>(cols));
        for (int i = 0; i < rows; ++i)
            for (int j = 0; j < cols; ++j)
                f >> blk[i][j];
    };
    read_block2d(r.tx);
    read_block2d(r.wtx);
    int k; f >> k;
    r.evals.resize(k);
    for (auto& v : r.evals) f >> v;
    return r;
}

// ---------------------------------------------------------------------------
// Reporting helpers
// ---------------------------------------------------------------------------
static int g_pass = 0, g_fail = 0;

static bool chk(const std::string& name, bool ok, const std::string& detail = "") {
    std::cout << "  [" << (ok ? "PASS" : "FAIL") << "] " << name;
    if (!detail.empty()) std::cout << "  (" << detail << ")";
    std::cout << "\n";
    ok ? ++g_pass : ++g_fail;
    return ok;
}

static void hdr(const std::string& title) {
    std::cout << "\n" << std::string(64, '=') << "\n"
              << "  " << title << "\n"
              << std::string(64, '=') << "\n";
}

static double max_abs_diff(const std::vector<std::vector<double>>& a,
                           const std::vector<std::vector<double>>& b) {
    double m = 0.0;
    for (std::size_t i = 0; i < a.size(); ++i)
        for (std::size_t j = 0; j < a[i].size(); ++j)
            m = std::max(m, std::abs(a[i][j] - b[i][j]));
    return m;
}

// ---------------------------------------------------------------------------
int main() {
    // ---- load reference data -----------------------------------------------
    Ref ref;
    try {
        ref = read_ref("lcg_full_ref.txt");
    } catch (const std::exception& e) {
        std::cerr << "ERROR: " << e.what() << "\n";
        return 1;
    }

    // ---- generate and center data ------------------------------------------
    auto X = gen_data();
    center_inplace(X);

    // ---- 1. Loaded-params transform -----------------------------------------
    hdr("1. Loaded-params transform (vs Python reference)");
    {
        auto p = full_pca_params::make_pca();

        std::vector<std::vector<double>> tx_cpp(REF_NTX);
        for (int i = 0; i < REF_NTX; ++i)
            tx_cpp[i] = p.transform_one(X[i]);

        double diff = max_abs_diff(tx_cpp, ref.tx);
        chk("Normal transform  abs_err < 1e-10", diff < 1e-10,
            "max_abs=" + std::to_string(diff));

        // whitened
        std::vector<std::vector<double>> wtx_cpp(REF_NTX);
        for (int i = 0; i < REF_NTX; ++i)
            wtx_cpp[i] = p.transform_one(X[i], /*whiten=*/true);

        double wdiff = max_abs_diff(wtx_cpp, ref.wtx);
        chk("Whitened transform  abs_err < 1e-10", wdiff < 1e-10,
            "max_abs=" + std::to_string(wdiff));
    }

    // ---- 2. transform_one vs transform_one_raw consistency -----------------
    hdr("2. transform_one vs transform_one_raw consistency");
    {
        auto p = full_pca_params::make_pca();
        auto v1 = p.transform_one(X[0]);
        auto v2 = p.transform_one_raw(X[0].data(), X[0].size());

        bool same = (v1.size() == v2.size());
        for (std::size_t j = 0; same && j < v1.size(); ++j)
            same = (v1[j] == v2[j]);
        chk("transform_one == transform_one_raw (bit-exact)", same);

        // whitened
        auto w1 = p.transform_one(X[0], true);
        auto w2 = p.transform_one_raw(X[0].data(), X[0].size(), true);
        bool wsame = (w1.size() == w2.size());
        for (std::size_t j = 0; wsame && j < w1.size(); ++j)
            wsame = (w1[j] == w2[j]);
        chk("whitened: transform_one == transform_one_raw (bit-exact)", wsame);
    }

    // ---- 3. transform() delegates to transform_one() ----------------------
    hdr("3. transform() == per-row transform_one()");
    {
        auto p = full_pca_params::make_pca();
        auto batch = p.transform(
            std::vector<std::vector<double>>(X.begin(), X.begin() + REF_NTX));

        double diff = 0.0;
        for (int i = 0; i < REF_NTX; ++i) {
            auto row = p.transform_one(X[i]);
            for (int j = 0; j < REF_K; ++j)
                diff = std::max(diff, std::abs(batch[i][j] - row[j]));
        }
        chk("transform() matches per-row transform_one()  diff==0", diff == 0.0,
            "max_abs=" + std::to_string(diff));
    }

    // ---- 4. C++ Jacobi fit eigenvalues vs Python reference -----------------
    hdr("4. C++ Jacobi fit eigenvalues vs Python eigh");
    {
        pca::PCA<double> p(REF_K);
        p.fit(X);
        const auto& ev_cpp = p.explained_variance_values();

        double max_rel = 0.0;
        for (int k = 0; k < REF_K; ++k) {
            double denom = std::abs(ref.evals[k]) + 1e-300;
            max_rel = std::max(max_rel, std::abs(ev_cpp[k] - ref.evals[k]) / denom);
        }
        chk("Jacobi eigenvalues  rel_err < 1e-6", max_rel < 1e-6,
            "max_rel=" + std::to_string(max_rel));
    }

    // ---- 5. Whitening output variance --------------------------------------
    hdr("5. Whitening output variance (full dataset)");
    {
        auto p = full_pca_params::make_pca();

        // Compute whitened transform of all N samples
        const int N = GEN_N;
        const int K = REF_K;
        std::vector<std::vector<double>> wtx(N);
        for (int i = 0; i < N; ++i)
            wtx[i] = p.transform_one(X[i], /*whiten=*/true);

        // Sample variance (ddof=1) of each output dimension should be ≈ 1.0
        double max_var_err = 0.0;
        for (int k = 0; k < K; ++k) {
            double sum = 0.0, sum2 = 0.0;
            for (int i = 0; i < N; ++i) {
                sum  += wtx[i][k];
                sum2 += wtx[i][k] * wtx[i][k];
            }
            double mean = sum / N;
            double var  = (sum2 - N * mean * mean) / (N - 1);   // ddof=1
            max_var_err = std::max(max_var_err, std::abs(var - 1.0));
        }
        chk("Whitened sample-var ≈ 1.0  abs_err < 1e-10", max_var_err < 1e-10,
            "max_abs=" + std::to_string(max_var_err));
    }

    // ---- 6. n_components clamping ------------------------------------------
    hdr("6. n_components clamping (K > D)");
    {
        pca::PCA<double> p(GEN_D + 999);  // ask for more than D features
        p.fit(X);
        bool clamped = (p.n_features() == GEN_D);
        chk("k_ clamped to D when n_components > D", clamped,
            "n_features=" + std::to_string(p.n_features()));
    }

    // ---- 7. Error handling --------------------------------------------------
    hdr("7. Error handling");
    {
        // transform before fit
        try {
            pca::PCA<double> p(5);
            p.transform(std::vector<std::vector<double>>(1, std::vector<double>(GEN_D)));
            chk("transform before fit throws", false);
        } catch (const std::runtime_error&) {
            chk("transform before fit throws", true);
        }

        // transform_one wrong size
        try {
            auto p = full_pca_params::make_pca();
            p.transform_one(std::vector<double>(GEN_D + 1, 0.0));
            chk("transform_one wrong size throws", false);
        } catch (const std::invalid_argument&) {
            chk("transform_one wrong size throws", true);
        }

        // transform_one_raw wrong size
        try {
            auto p = full_pca_params::make_pca();
            std::vector<double> buf(GEN_D + 1, 0.0);
            p.transform_one_raw(buf.data(), buf.size());
            chk("transform_one_raw wrong size throws", false);
        } catch (const std::invalid_argument&) {
            chk("transform_one_raw wrong size throws", true);
        }

        // load with bad components size
        try {
            pca::PCA<double> p(5);
            p.load(10, 5,
                   std::vector<double>(99),   // should be 50
                   std::vector<double>(5),
                   std::vector<double>(5),
                   1.0);
            chk("load bad components size throws", false);
        } catch (const std::invalid_argument&) {
            chk("load bad components size throws", true);
        }

        // load with K <= 0
        try {
            pca::PCA<double> p(5);
            p.load(0, 5,
                   std::vector<double>(0),
                   std::vector<double>(5),
                   std::vector<double>(5),
                   1.0);
            chk("load D=0 throws", false);
        } catch (const std::invalid_argument&) {
            chk("load D=0 throws", true);
        }

        // PCA with n_components <= 0
        try {
            pca::PCA<double> p(0);
            chk("PCA(0) throws", false);
        } catch (const std::invalid_argument&) {
            chk("PCA(0) throws", true);
        }
    }

    // ---- 8. float type (smoke test) ----------------------------------------
    hdr("8. float type smoke test");
    {
        pca::PCA<float> fp(REF_K);
        std::vector<std::vector<float>> Xf(GEN_N, std::vector<float>(GEN_D));
        for (int i = 0; i < GEN_N; ++i)
            for (int j = 0; j < GEN_D; ++j)
                Xf[i][j] = static_cast<float>(X[i][j]);
        fp.fit(Xf);

        const auto& ev_f = fp.explained_variance_values();
        double max_rel = 0.0;
        for (int k = 0; k < REF_K; ++k) {
            double denom = std::abs(ref.evals[k]) + 1e-300;
            max_rel = std::max(max_rel, std::abs(static_cast<double>(ev_f[k]) - ref.evals[k]) / denom);
        }
        // float precision: expect rel_err < 1e-5
        chk("float PCA eigenvalues  rel_err < 1e-5", max_rel < 1e-5,
            "max_rel=" + std::to_string(max_rel));

        auto tx_f = fp.transform_one(Xf[0]);
        chk("float transform_one returns K values",
            static_cast<int>(tx_f.size()) == REF_K);
    }

    // ---- summary ------------------------------------------------------------
    hdr("Summary");
    std::cout << "  PASS: " << g_pass << "   FAIL: " << g_fail << "\n";
    std::cout << "  Result: " << (g_fail == 0 ? "ALL PASS" : "SOME FAILURES") << "\n";
    return g_fail > 0 ? 1 : 0;
}

File: test_full_audit.py
"""
Full audit: custom PCA vs sklearn, then export reference data for C++ audit.

Settings: SEED=143, N=3000, D=100, K=20, N_TX=50
"""
import sys
import numpy as np
from pca import PCA

SEED  = 143
N     = 3000
D     = 100
K     = 20
N_TX  = 50

# ---------------------------------------------------------------------------
def lcg(seed: int) -> int:
    return (1664525 * seed + 1013904223) & 0xFFFF_FFFF

def gen_lcg(seed: int, n: int, d: int):
    out = np.empty(n * d, dtype=np.float64)
    for i in range(n * d):
        seed = lcg(seed)
        out[i] = seed / 4294967296.0
    return out.reshape(n, d), seed

def hdr(title: str):
    print(f"\n{'='*64}")
    print(f"  {title}")
    print(f"{'='*64}")

_all_pass = True

def check(name: str, ok: bool, detail: str = "") -> bool:
    global _all_pass
    status = "PASS" if ok else "FAIL"
    print(f"  [{status}] {name}" + (f"  ({detail})" if detail else ""))
    if not ok:
        _all_pass = False
    return ok

# ---------------------------------------------------------------------------
def main():
    from sklearn.decomposition import PCA as SkPCA

    # ---- data ---------------------------------------------------------------
    X_raw, _ = gen_lcg(SEED, N, D)
    col_mean  = X_raw.mean(axis=0)           # (D,)
    X         = X_raw - col_mean             # centered, what we pass to our PCA

    # ---- fit ----------------------------------------------------------------
    my_pca = PCA(K)
    my_pca.fit(X)

    sk_pca = SkPCA(n_components=K, svd_solver="full")
    sk_pca.fit(X_raw)                        # sklearn centers internally

    hdr("1. Eigenvalues / explained variance")

    my_ev = my_pca.explained_variance_values()   # (K,)
    sk_ev = sk_pca.explained_variance_           # (K,)
    rel   = np.abs(my_ev - sk_ev) / (np.abs(sk_ev) + 1e-300)
    check("Eigenvalues  rel_err < 1e-10", rel.max() < 1e-10,
          f"max_rel={rel.max():.3e}")

    my_er = my_pca.explained_variance_ratio()    # (K,)
    sk_er = sk_pca.explained_variance_ratio_     # (K,)
    check("Explained ratio abs_err < 1e-10",
          np.abs(my_er - sk_er).max() < 1e-10,
          f"max_abs={np.abs(my_er - sk_er).max():.3e}")

    print(f"  eigenvalues top-5 (ours):   {my_ev[:5]}")
    print(f"  eigenvalues top-5 (sklearn): {sk_ev[:5]}")
    print(f"  explained ratio sum: {my_er.sum():.8f}")

    # ---- components (sign-aligned) -----------------------------------------
    hdr("2. Principal components")

    # sklearn.components_ shape (K, D); ours (D, K)
    my_comp = my_pca.components_        # (D, K)
    sk_comp = sk_pca.components_.T      # (D, K)

    # sign convention: align each column of our components to sklearn's
    signs = np.sign(np.einsum("dk,dk->k", my_comp, sk_comp))
    signs[signs == 0] = 1.0
    my_aligned = my_comp * signs        # (D, K)

    abs_comp = np.abs(my_aligned - sk_comp).max()
    check("Components  abs_err < 1e-10", abs_comp < 1e-10,
          f"max_abs={abs_comp:.3e}")

    # ---- transform ----------------------------------------------------------
    hdr("3. Transform (N_TX=%d samples)" % N_TX)

    X_tx     = X[:N_TX]                 # centered
    my_tx    = my_pca.transform(X_tx)   # (N_TX, K)
    sk_tx    = sk_pca.transform(X_raw[:N_TX])  # (N_TX, K)

    # sign-align transform columns (same signs as components)
    my_tx_al = my_tx * signs            # (N_TX, K)
    abs_tx   = np.abs(my_tx_al - sk_tx).max()
    check("Transform  abs_err < 1e-10", abs_tx < 1e-10,
          f"max_abs={abs_tx:.3e}")

    # ---- whitened transform -------------------------------------------------
    hdr("4. Whitened transform (N_TX samples)")

    my_wtx   = my_pca.transform(X_tx, whiten=True)   # (N_TX, K)
    sk_wtx   = sk_pca.transform(X_raw[:N_TX]) / np.sqrt(sk_pca.explained_variance_)
    my_wtx_al = my_wtx * signs
    abs_wtx  = np.abs(my_wtx_al - sk_wtx).max()
    check("Whitened transform  abs_err < 1e-10", abs_wtx < 1e-10,
          f"max_abs={abs_wtx:.3e}")

    # ---- whitening variance on full dataset ---------------------------------
    hdr("5. Whitening output variance")

    wtx_full = my_pca.transform(X, whiten=True)   # (N, K)
    # np.var uses N in denominator; eigenvalues use N-1 → expect ~(N-1)/N
    var_ddof0    = np.var(wtx_full, axis=0)
    expected_var = (N - 1) / N
    abs_wvar     = np.abs(var_ddof0 - expected_var).max()
    check("Whitened var ≈ (N-1)/N  abs_err < 1e-10", abs_wvar < 1e-10,
          f"max_abs={abs_wvar:.3e}, expected≈{expected_var:.7f}")

    # sample variance (ddof=1) should be ≈ 1.0
    var_ddof1 = np.var(wtx_full, axis=0, ddof=1)
    abs_wvar1 = np.abs(var_ddof1 - 1.0).max()
    check("Whitened sample-var ≈ 1.0  abs_err < 1e-10", abs_wvar1 < 1e-10,
          f"max_abs={abs_wvar1:.3e}")

    # ---- error handling -----------------------------------------------------
    hdr("6. Error handling")

    unfitted = PCA(K)
    try:
        unfitted.transform(X[:5])
        check("transform before fit raises", False)
    except RuntimeError:
        check("transform before fit raises", True)

    try:
        unfitted.explained_variance_ratio()
        check("explained_variance_ratio before fit raises", False)
    except RuntimeError:
        check("explained_variance_ratio before fit raises", True)

    try:
        my_pca.transform(X[:5, :D//2])     # wrong D
        check("transform wrong-D raises", False)
    except ValueError:
        check("transform wrong-D raises", True)

    try:
        PCA(-1)
        check("PCA(n_components=-1) raises", False)
    except ValueError:
        check("PCA(n_components=-1) raises", True)

    # ---- export -------------------------------------------------------------
    hdr("7. Export for C++ audit")

    my_pca.export_to_hpp("lcg_full_pca_params.hpp", namespace="full_pca_params")
    print("  Exported lcg_full_pca_params.hpp")

    with open("lcg_full_ref.txt", "w") as f:
        # block 1: normal transforms
        f.write(f"{N_TX} {K}\n")
        for row in my_tx:
            f.write(" ".join(f"{v:.17g}" for v in row) + "\n")
        # block 2: whitened transforms
        f.write(f"{N_TX} {K}\n")
        for row in my_wtx:
            f.write(" ".join(f"{v:.17g}" for v in row) + "\n")
        # block 3: eigenvalues (for C++ Jacobi comparison)
        f.write(f"{K}\n")
        f.write(" ".join(f"{v:.17g}" for v in my_ev) + "\n")
    print("  Exported lcg_full_ref.txt")

    # ---- final --------------------------------------------------------------
    hdr("Summary")
    print(f"  Result: {'ALL PASS' if _all_pass else 'SOME FAILURES'}")
    if not _all_pass:
        sys.exit(1)


if __name__ == "__main__":
    main()
